pkg/authn/challenge/challenge.go

```diff
@@ -0,0 +1,118 @@
+package challenge
+
+import (
+	"context"
+	"crypto/rsa"
+	"embed"
+	"encoding/json"
+	"net/http"
+
+	"github.com/bacalhau-project/bacalhau/pkg/authn"
+	"github.com/bacalhau-project/bacalhau/pkg/lib/policy"
+	"github.com/bacalhau-project/bacalhau/pkg/system"
+	"github.com/lestrrat-go/jwx/jwk"
+	"github.com/samber/lo"
+)
+
+//go:embed *.rego
+var policies embed.FS
+
+type policyData struct {
+	SigningKey jwk.Key `json:"signingKey"`
+	NodeID     string  `json:"nodeId"`
+	ClientID   string  `json:"clientId"`
+}
+
+type response struct {
+	PhraseSignature string `json:"phraseSignature"`
+
+	// A base64 encoded PKCS1 public key, as generated by config.encodePublicKey
+	PublicKey string `json:"publicKey"`
+}
+
+type request struct {
+	InputPhrase []byte `json:"inputPhrase"`
+}
+
+type phraser = func() ([]byte, error)
+
+type challengeAuthenticator struct {
+	authnPolicy *policy.Policy
+	getPhrase   phraser
+	key         jwk.Key
+	nodeID      string
+
+	challenge policy.Query[policyData]
+}
+
+func NewAuthenticator(p *policy.Policy, getPhrase phraser, key *rsa.PrivateKey, nodeID string) authn.Authenticator {
+	return challengeAuthenticator{
+		authnPolicy: p,
+		getPhrase:   getPhrase,
+		key:         lo.Must(jwk.New(key)),
+		nodeID:      nodeID,
+		challenge:   policy.AddQuery[policyData](p, authn.PolicyTokenRule),
+	}
+}
+
+// Authenticate implements Authenticator.
```

Comment:
Is it better to use an interface guard rather than a comment for this (especially as the comment is confusing and might be best as // challengeAuthenticator implements Authenticator

```
var _ Authenticator = (*challengeAuthenticator)(nil)
```

Reply:
The interface is actually already guarded because `NewAuthenticator` returns the interface and not the type itself, so any interface mismatch would show up in the constructor. These comments are auto-generated by VS Code when you ask it to create method stubs for an interface. I'll just remove them.